### PR TITLE
doc: Add ACRN secureboot setup with GRUB

### PR DESCRIPTION
**Pull request by yonghuah**
Add guide to enable grub secure boot on ACRN.

Signed-off-by: Yonghua Huang <yonghua.huang@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by yonghuah** [comment]
@NanlinXie  @dbkinder  please review this PR ,thanks.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+This document shows how to enable ACRN secure boot with GRUB including:
+
+  -  ACRN Secure Boot Sequence
+  -  Generate GPG Key
+  -  Setup Standalone GRUB EFI Binary
+  -  UEFI Secure Boot
+
+  Validation Environment:
+	Hardware Platform: INTEL WHL
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:14]: Can you change this to a pointer to the platform from the "Supported Hardware" list?

If we expect that this method can work with other platforms too, it would be good to add a note in that sense.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	ACRN Scenario: Industry
+	Service VM: Ubuntu 18.04
+	GRUB: 2.02
+
+Note: Please make sure you have alternative ways to restore GRUB when you got stuck.
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:19]: I would rephrase this into something like: "Note that Grub may stop booting in case of problems, make sure you know how to recover a bootloader on your platform". @dbkinder , any better suggestion / (Not quite happy with what I wrote)

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+     .. image:: images/acrn-secureboot-flow.png
+        :align: center
+        :width: 600px
+
+
+  Refer to :ref:`using_grub` for details to enable GRUB on ACRN.
+
+  From point view of secureboot:
+	- UEFI firmware verifies Shim/GRUB
+	- GRUB verifies ACRN, Service VM kernel and Prelaunch VM kernel
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:35]:
```suggestion
	- GRUB verifies ACRN, Service VM kernel and pre-launched User VM kernel
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	- SOS kernel verifies DM and UOS OVMF thru ACRN DM
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:36]:
```suggestion
	- Service VM OS kernel verifies the Device Model (`acrn-dm`) and User VM OVMF bootloader (with the help of `acrn-dm`)
```

Would this still be correct?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	- UOS OVMF starts the guest side verified boot process
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:37]:
```suggestion
	- User VM OVMF starts the guest side verified boot process
```

I'm wondering if it wouldn't be better to talk about the "virtual bootloader" instead of "User VM OVMF". @dbkinder , thoughts?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  This document provides guide to enable verification function of GRUB to verify ACRN binaries(such acrn.bin, service VM bzImage and Prelaunch VM kernel image if it is present).
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:39]:
```suggestion
  This document provides guide to enable verification function of GRUB to verify ACRN binaries (such `acrn.bin`, Service VM kernel (`bzImage`) and pre-launched User VM kernel image if it is present).
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  -  ACRN Secure Boot Sequence
+  -  Generate GPG Key
+  -  Setup Standalone GRUB EFI Binary
+  -  UEFI Secure Boot
+
+  Validation Environment:
+	Hardware Platform: INTEL WHL
+	ACRN Scenario: Industry
+	Service VM: Ubuntu 18.04
+	GRUB: 2.02
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:17]: We use 2.04 with a patch in our main Getting Started Guide... can we use that one too here? I would prefer that we stick to a single version if at all possible.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+Note: Please make sure you have alternative ways to restore GRUB when you got stuck.
+
+ACRN Secure Boot Sequence
+*************************
+
+  ACRN can be booted by Multiboot compatible bootloader, following diagram illustrates the boot sequence of ACRN with GRUB:
+
+     .. image:: images/acrn-secureboot-flow.png
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:26]: Can you update the diagram to use "Service VM kernel" instead of "SOS kernel"?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+  This document provides guide to enable verification function of GRUB to verify ACRN binaries(such acrn.bin, service VM bzImage and Prelaunch VM kernel image if it is present).
+
+
+Generate GPG Key
+****************
+
+  GRUB supports loading of GPG signed files only if digital signatures enabled.
+  Example to generate GPG signing key::
+  $mkdir --mode 0700 keys
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:47]:
```suggestion
  $ mkdir --mode 0700 keys
```
Do this too for the lines below.

I suspect this code-block formatting is not rendered as expected, have you verified the final rendering (using `make doc`)?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+  ``gpg --gen-key`` generates key pair, private key will be used to sign GRUB configuration files and ACRN binaries, public key will be embedded in GRUB and is used to verify GRUB configuration files or binaries GRUB trys to load.
+
+
+Setup Standalone GRUB EFI Binary
+********************************
+
+Prepare Initial GRUB Configuration grub.init.cfg
+================================================
+  Store the following minimal grub config in grub.init.cfg, ``set check_signatures =enforce`` is to enable digital signatures in GRUB::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:59]:
```suggestion
  Store the following minimal grub config in grub.init.cfg, ``set check_signatures=enforce`` is to enable digital signatures in GRUB::
```

Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:59]: I think @dbkinder prefers that we use :file:`grub.init.cfg` for files.

Where is that file located by the way?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+	set check_signatures=enforce
+	export check_signatures
+
+	search --no-floppy --fs-uuid --set=root ESP_UUID
+	configfile /grub.cfg
+	echo /grub.cfg did not boot the system, rebooting in 10 seconds.
+	sleep 10
+	reboot
+
+  Replace the ESP_UUID with UUID of your EFI system partition, with the following output of ``lsblk -f``, replace it with ``24FC-BE7A``.::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:69]: @dbkinder can you help reword this sentence?

Also, just wondered where we should replace the double-backticks around `lbblk` by :command:`lsblk -f`?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+		--pubkey ./boot.key \
+		--output ./grubx64.efi  \
+		"boot/grub/grub.cfg=./grub.init.cfg" \
+		"boot/grub/grub.cfg.sig=./grub.init.cfg.sig"
+
+	echo "writing signed grub.efi to '$TARGET_EFI'"
+	sudo cp ./grubx64.efi "$TARGET_EFI"
+
+
+  Note: You should backup current GRUB image in '$TARGET_EFI', then you can restore your grub in case you got stuck after updating it.
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:126]: Replace with `.. note::`

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 2 above)
Comment (gvancuts) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:19]: Use `.. note::` here for proper rendering of the note.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+	 echo 'loading ACRN...'
+	 multiboot2 /boot/acrn/acrn.bin  root=PARTUUID="c8ee7d92-8935-4e86-9e12-05dbeb412ad6"
+	 module2 /boot/bzImage Linux_bzImage
+	}
+
+  Update the UUID (``--set``) and PARTUUID (``root=`` parameter) of the root partition (e.g. /dev/nvme0n1p2) according to
+  your your hardware(e.g. use ``blkid`` tool).
+
+  Copy grub.cfg to root of ESP(e.g. /boot/efi/EFI/).
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:153]:
```suggestion
  Copy :file:`grub.cfg` to your ESP (e.g. `/boot/efi/EFI/`).
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+
+Sign grub.cfg and ACRN Binaries
+===============================
+
+  grub.cfg file and all ACRN binaries that will be loaded by GRUB MUST be signed with GPG key.
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:159]:
```suggestion
  The :file:`grub.cfg` file and all ACRN binaries that will be loaded by GRUB **must** be signed with GPG key.
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+  Now, you can reboot and system will boot with single GRUB EFI binary.
+  GRUB will refuse to boot if any files to be loaded are tampered.
+
+
+UEFI Secure Boot
+================
+
+  UEFI secure boot can be enabled to verify GRUB EFI binary, which may be verified by SHIM or UEFI firmware (without SHIM).
+  Refer to `How to sign things for Secure Boot <https://ubuntu.com/blog/how-to-sign-things-for-secure-boot>`_
+  for detailed steps(not verified).
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:176]:
```suggestion
  for detailed steps (not verified).
```

**Review by gvancuts — CHANGES_REQUESTED** [review]
Nice to see this new document!! :-)

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 7 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:17]: @gvancuts  i tested using 2.02,  QA team will re-try these steps with 2.04, i can update it after that.  make sense?

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 9 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:47]: @gvancuts my local make doc does not work,  not see error reported from CI /build doc, are they same?

//my local build doc err, any hint?

make[1]: Entering directory '/home/yh/workspace/acrn/acrn-hypervisor/doc'
Makefile:60: recipe for target 'html' failed
make[1]: [html] Error 2 (ignored)

New errors/warnings found, please fix them:
==============================================


Extension error:
Could not import extension sphinx.builders.applehelp (exception: cannot import name 'RemovedInSphinx40Warning')
Makefile:60: recipe for target 'html' failed
make[1]: *** [html] Error 1
make[1]: Leaving directory '/home/yh/workspace/acrn/acrn-hypervisor/doc'
Makefile:184: recipe for target 'doc' failed
make: *** [doc] Error 2

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 10 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:59]: @gvancuts  it is a tmp file and will be built into grub finally.

**Comment by yonghuah** [comment]
@gvancuts  @dbkinder  have update this doc, please review.  many thanks for so helpful comments.

-Yonghua

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 7 above)
Comment (gvancuts) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:17]: > @gvancuts i tested using 2.02, QA team will re-try these steps with 2.04, i can update it after that. make sense?

Do you know when they will test this? I would like them to test the version 2.04 **with** the patch we use in our GSG too. Can they do that?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  -  Setup Standalone GRUB EFI Binary
+  -  UEFI Secure Boot
+
+  Validation Environment:
+	Hardware Platform: WHL-IPC-I7, Supported hardware described in :ref:`hardware`.
+	ACRN Scenario: Industry
+	Service VM: Ubuntu 18.04
+	GRUB: 2.02
+
+..note::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:19]: The formatting seems off, it should be something like:
```
.. note::
   Grub may stop... blah blah
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+        :align: center
+        :width: 600px
+
+
+  Refer to :ref:`using_grub` for details to enable GRUB on ACRN.
+
+  From point view of secureboot:
+	- UEFI firmware verifies Shim/GRUB
+	- GRUB verifies ACRN, Service VM kernel and pre-launched User VM kernel
+	- Service VM OS kernel verifies the Device Model (`acrn-dm`) and User VM OVMF bootloader (with the help of `acrn-dm`)
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:37]:
```suggestion
	- Service VM OS kernel verifies the Device Model (``acrn-dm``) and User VM OVMF bootloader (with the help of ``acrn-dm``)
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+	- User VM virtual bootloader (e.g. OVMF) starts the guest side verified boot process
+
+   This document provides guide to enable verification function of GRUB to verify ACRN binaries (such `acrn.bin`, Service VM kernel (`bzImage`) and pre-launched User VM kernel image if it is present).
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:40]:
```suggestion
   This document provides guide to enable verification function of GRUB to verify ACRN binaries (such ``acrn.bin``, Service VM kernel (``bzImage``) and pre-launched User VM kernel image if it is present).
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+
+Generate GPG Key
+****************
+
+  GRUB supports loading of GPG signed files only if digital signatures enabled.
+  Example to generate GPG signing key::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:47]: I _think_ you need a blank line here just before the code-block. Do you run `make doc` after your updates?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+  ``gpg --gen-key`` generates key pair, private key will be used to sign GRUB configuration files and ACRN binaries, public key will be embedded in GRUB and is used to verify GRUB configuration files or binaries GRUB trys to load.
+
+
+Setup Standalone GRUB EFI Binary
+********************************
+
+Prepare Initial GRUB Configuration grub.init.cfg
+================================================
+  Create file ``grub.init.cfg`` to store the following minimal grub configuration, ``set check_signatures=enforce`` is to enable digital signatures in GRUB::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:60]: @dbkinder can provide more details but ReST is sensitive to indentation, I'm wondering what it does when it sees that sentence starting with three spaces.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+		--pubkey ./boot.key \
+		--output ./grubx64.efi  \
+		"boot/grub/grub.cfg=./grub.init.cfg" \
+		"boot/grub/grub.cfg.sig=./grub.init.cfg.sig"
+
+	echo "writing signed grub.efi to '$TARGET_EFI'"
+	sudo cp ./grubx64.efi "$TARGET_EFI"
+
+
+..note::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:127]: Same here, indentation is incorrect.

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+
+  Copy :file:`grub.cfg` to your ESP (e.g. `/boot/efi/EFI/`).
+
+
+Sign grub.cfg and ACRN Binaries
+===============================
+
+  The :file:`grub.cfg` file and all ACRN binaries that will be loaded by GRUB **must** be signed with GPG key.
+
+  Sign script example::
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:163]: Better call that a "sign sequence example". It cannot really be copy-pasted into a file and ran as a script

**Comment by gvancuts** [comment]
Not sure if you've resolved your `make doc` error, but I would first check that you have correctly installed all documentation generation tools correctly: https://projectacrn.github.io/latest/tutorials/docbuild.html#installing-the-documentation-tools

Your error looks different to what was reported above in this thread.

**Comment by dbkinder** [comment]
@gvancuts @yonghuah I went through the document and fixed the reST formatting issues, as well as some writing cleanup.  I pushed the changes back directly to @yonghuah repo, so if @gvancuts approves, I'm ready to merge.  (Only comment not addressed was whether we use grub 2.02 or 2.04)

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Review by gvancuts — APPROVED** [review]
I believe `grub` is typically spelled `GRUB` (capital letters). I have spotted a few minor other things to fix but otherwise this looks good to me!

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+enabled. Here's an example of generating a GPG signing key::
+
+  mkdir --mode 0700 keys
+  gpg --homedir keys  --gen-key
+  gpg --homedir keys --export > boot.key
+
+The ``gpg --gen-key`` command generates a public and private key pair.
+The private key is used to sign grub configuration files and ACRN
+binaries. The public key will be embedded in grub and is used to verify
+grub configuration files or binaries grub trys to load.
```
[the remark this answers is not in the document]
Comment (dbkinder) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:65]: oops :)

**Comment by yonghuah** [comment]
@dbkinder  Please hold on this PR, i'm revising:
1. address review comments.
2. one more section is required about enable uefi secure boot part ,which depends on a poc testing, i will update this PR today.

Thank you for all review comment and sorry for late response.

**Comment by yonghuah** [comment]
@dbkinder @gvancuts  Please hold on this PR , will update today, thanks.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by dbkinder** [comment]
OK.  I mad come changes directly to your repo branch so you can just add some more and amend your commit.
-- david

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 7 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst]: we have tested with grub 2.04, thanks

**Comment by yonghuah** [comment]
@dbkinder  have updated this PR, please review, thanks.

-Yonghua

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 24]
```diff
@@ -0,0 +1,256 @@
...
+    echo "writing signed grub.efi to '$TARGET_EFI'"
+    sudo cp ./grubx64.efi "$TARGET_EFI"
+
+
+.. rst-class:: numbered-step
+
+Prepare grub.cfg
+****************
+
+Define the menu entry for your system in a new GRUB configuration :file: `grub.cfg`.
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:169]: Wouldn't that file be overwritten by someone doing a `sudo update-grub`?

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+-  Generate GPG Key
+-  Setup Standalone GRUB EFI Binary
+-  Enable UEFI Secure Boot
+
+**Validation Environment:**
+
+- Hardware Platform: WHL-IPC-I7, Supported hardware described in
+  :ref:`hardware`.
+- ACRN Scenario: Industry
+- Service VM: Yocto
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:18]: This mentions Yocto, although I believe it mentionned Ubuntu previously.

The script below still seems to be for Ubuntu, e.g. `TARGET_EFI='/boot/efi/EFI/ubuntu/grubx64.efi'`

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+Replace the ``GRUB_PASSWORD_HASH`` with the result of the :command:`grub-mkpasswd-pbkdf2`
+with your custom passphrase.
+
+Use this command to sign the :file:`grub.init.cfg` file with your private
+GPG key and create the :file:`grub.init.cfg.sig`::
+
+    gpg --homedir keys --detach-sign grub.init.cfg
+
+
+Create a single GRUB EFI Binary
```
Comment (dbkinder) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:123]: I'm thinking this should be ``signed`` and not ``single`` here, and on line 126 below

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+
+    sbsign --key db.key --cert db.crt path/to/grubx64.efi
+
+:file:`grubx64.efi.signed` will be created, it will be your bootloader.
+
+Enroll UEFI Keys To UEFI Firmware
+=================================
+
+Enroll ``PK``(:file:`PK.der`), ``KEK``(:file:`KEK.der`) and ``db``(:file:`db.der`) in
+Secure Boot Configuration UI, which depends on your platform UEFI firmware. From now on,
```
Comment (dbkinder) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:254]: We should include an example of how to "Enroll" these files on one of our supported platforms, for example, the one you validated listed at the top of the page.

**Review by dbkinder — CHANGES_REQUESTED** [review]
A few more comments to address...

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 26 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:123]: @dbkinder, not signed, suppose "standalone" is better, right?

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 24 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:169]: Will not, path/to/this/grub.cfg will not be touched by 'update-grub', unless user intend to do so, e.g. put it under /boot/grub/

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 25 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst]: @gvancuts , tested on both yoctor and ubuntu, have update this mismatch, thanks.

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 27 above)
Comment (yonghuah) [reply doc/tutorials/acrn-secure-boot-with-grub.rst]: @dbkinder  have updated this part, please review, thank you.
-Yonghua

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/tutorials/acrn-secure-boot-with-grub.rst [hunk 28]
```diff
@@ -0,0 +1,259 @@
...
+    cert-to-efi-sig-list db.crt db.esl
+
+-Sign ESL files::
+
+    sign-efi-sig-list -k PK.key -c PK.crt PK PK.esl PK.auth
+    sign-efi-sig-list -k PK.key -c PK.crt KEK KEK.esl KEK.auth
+    sign-efi-sig-list -k KEK.key -c KEK.crt db db.esl db.auth
+
+The keys to be enrolled in UEFI firmware: :file:`PK.der` , :file:`KEK.der` , :file:`db.der`.
+The keys to sign bootloader image :file:'grubx64.efi': :file:`db.key` , :file:`db.crt`.
```
Comment (gvancuts) [inline doc/tutorials/acrn-secure-boot-with-grub.rst:241]:
```suggestion
The keys to sign bootloader image :file:`grubx64.efi`: :file:`db.key` , :file:`db.crt`.
```

File: doc/tutorials/acrn-secure-boot-with-grub.rst (hunk 24 above)
Comment (gvancuts) [reply doc/tutorials/acrn-secure-boot-with-grub.rst:169]: OK, thanks!

**Review by gvancuts — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
I fixed a few formatting and linking issues, LGTM. Thanks for the update!

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>
